src/software/util/design_patterns/generic_factory.h

```diff
@@ -0,0 +1,117 @@
+#pragma once
+#include <memory>
+#include <string>
+#include <vector>
+#include <unordered_map>
+#include <functional>
+//#include "software/backend/backend.h" // Should make generic.h
+
+// A quality of life typedef to make things shorter and more readable
+template <class IndexType, class TypeToCreate>
+using GenericRegistry = std::unordered_map<IndexType, std::function<std::unique_ptr<TypeToCreate>()>>;
+/**
+ * The GenericFactory is an Abstract class that provides an interface for Generic type
+ * Factories to follow. This makes it easy to maintain a list of factories and get the
+ * corresponding generic types through the generic interface.
+ */
+template <class IndexType, class TypeToCreate>
+class GenericFactory
+{
+public:
+    /**
+     * Returns a unique pointer to a newly constructed Generic of the given type/name
+     *
+     * @param generic_name The name of the Generic to construct. This value must be in the
+     * Generic registry
+     * @throws std::invalid_argument if the given generic_name is not found in the Generic
+     * registry
+     *
+     * @return a unique pointer to a newly constructed Generic of the given type/name
+     */
+    static std::unique_ptr<TypeToCreate> createGeneric(const std::string& generic_name);
+
+    /**
+     * Returns a const reference to the Generic registry. The registry is a map of Generic
+     * names to a "create" function that will create and return a unique_ptr to a new
+     * concrete instance of the Generic.
+     *
+     * @return a const reference to the Generic registry
+     */
+    static const GenericRegistry<IndexType, TypeToCreate>& getRegistry();
+
+    /**
+     * Returns a list of names of all the existing Generics
```

Comment:
Somewhat of a comment for this whole file. We're using the word "Generics" a lot but not really saying what that is. In different contexts / functions, is could be a type, object, or "creator" function. We should try use those words as descriptors instead of overusing the word "generic". The class already implies everything is generic, so I don't think we need to constantly repeat it.

Reply:
Addressed.